### PR TITLE
Include json files for fn source and sink

### PR DESCRIPTION
**Pull request by phanimarupaka**
@frankfarzan 

This PR is regarding the issue https://github.com/GoogleContainerTools/kpt/issues/647. Json files will be read as source(included in the PR) but will be written as yaml files to sink(current behavior). The logic here is to convert any json nodes to yaml nodes and process them as we are doing currently.

### CODE REVIEW

File: kyaml/kio/pkgio_reader.go [hunk 1]
```diff
@@ -162,10 +163,18 @@ type LocalPackageReader struct {
 
 	// SetAnnotations are annotations to set on the Resources as they are read.
 	SetAnnotations map[string]string `yaml:"setAnnotations,omitempty"`
+
+	// IncludeJSON indicates if the json resources files must be included during read
+	IncludeJSON bool
```
Comment (pwittrock) [inline kyaml/kio/pkgio_reader.go:168]: Don't add this field

File: cmd/config/internal/commands/source.go [hunk 2]
```diff
@@ -71,7 +71,7 @@ func (r *SourceRunner) runE(c *cobra.Command, args []string) error {
 
 	var inputs []kio.Reader
 	for _, a := range args {
-		inputs = append(inputs, kio.LocalPackageReader{PackagePath: a})
+		inputs = append(inputs, kio.LocalPackageReader{PackagePath: a, IncludeJSON: true})
```
Comment (pwittrock) [inline cmd/config/internal/commands/source.go:74]: Why not just change the `MatchFilesGlob`?

File: kyaml/kio/pkgio_reader.go [hunk 3]
```diff
@@ -162,10 +163,18 @@ type LocalPackageReader struct {
...
 }
 
 var _ Reader = LocalPackageReader{}
 
+const (
```
Comment (pwittrock) [inline kyaml/kio/pkgio_reader.go:173]: replace this with:

```go
var (
  MatchJSON = []string{"*.json"}
  MatchYAML = []string{"*.yaml", "*.yml"}
  MatchAll = append(MatchJSON, MatchYAML...)
)

```

**Comment by frankfarzan** [comment]
@phanimarupaka  As a follow up PR, consider sink preserving the file format using the `path` annotation. It's more intuitive and less surprising behavior.

File: kyaml/kio/pkgio_reader.go (hunk 1 above)
Comment (phanimarupaka) [reply kyaml/kio/pkgio_reader.go]: @pwittrock Done.

File: cmd/config/internal/commands/source.go (hunk 2 above)
Comment (phanimarupaka) [reply cmd/config/internal/commands/source.go]: Agree. This is simpler. Done.

File: kyaml/kio/pkgio_reader.go (hunk 3 above)
Comment (phanimarupaka) [reply kyaml/kio/pkgio_reader.go]: Done.

File: kyaml/kio/byteio_reader.go [hunk 4]
```diff
@@ -119,7 +122,14 @@ func (r *ByteReader) Read() ([]*yaml.RNode, error) {
 
 	index := 0
 	for i := range values {
-		decoder := yaml.NewDecoder(bytes.NewBufferString(values[i]))
+		value := values[i]
+		if strings.HasSuffix(r.Path, JSONSuffix) {
```
Comment (pwittrock) [inline kyaml/kio/byteio_reader.go:126]: We probably need to also update the package write to write json as well?

File: cmd/config/internal/commands/source.go [hunk 5]
```diff
@@ -71,7 +71,7 @@ func (r *SourceRunner) runE(c *cobra.Command, args []string) error {
 
 	var inputs []kio.Reader
 	for _, a := range args {
-		inputs = append(inputs, kio.LocalPackageReader{PackagePath: a})
```
Comment (pwittrock) [inline cmd/config/internal/commands/source.go:74]: Do we need to update the run command as well?  We probably want `source DIR | run`and `run DIR` to use the same file matching.

File: kyaml/kio/byteio_reader.go (hunk 4 above)
Comment (pwittrock) [inline kyaml/kio/byteio_reader.go:126]: Probably want to check all the JSONMatch values?

Comment (pwittrock) [reply kyaml/kio/byteio_reader.go:126]: Also, consider using filepath.Ext

File: kyaml/kio/byteio_reader.go [hunk 6]
```diff
@@ -101,6 +101,9 @@ type ByteReader struct {
 	// WrappingKind is set by Read(), and is the kind of the object that
 	// the read objects were originally wrapped in.
 	WrappingKind string
+
+	// Path indicates file path which is being read
+	Path string
```
Comment (pwittrock) [inline kyaml/kio/byteio_reader.go:106]: Probably want to use `SerializationMode` or something rather than the path.

File: cmd/config/internal/commands/source_test.go [hunk 7]
```diff
@@ -21,25 +21,37 @@ func TestSourceCommand(t *testing.T) {
...
-kind: Deployment
-metadata:
-  labels:
-    app: nginx2
-  name: foo
-  annotations:
-    app: nginx2
-spec:
-  replicas: 1
+	err = ioutil.WriteFile(filepath.Join(d, "f1.json"), []byte(`
```
Comment (pwittrock) [inline cmd/config/internal/commands/source_test.go:24]: Would you add an e2e test under `internal/commands/e2e` for source?

File: kyaml/yaml/types.go [hunk 8]
```diff
@@ -724,20 +724,29 @@ func (rn *RNode) UnmarshalJSON(b []byte) error {
...
+	node, err := Parse(yml)
 	if err != nil {
 		return nil, err
 	}
-	node, err := Parse(string(yml))
+	return node, nil
+}
+
+// ConvertJSONToYamlNode parses input json string and returns equivalent yaml string
+func ConvertJSONToYamlString(jsonStr string) (string, error) {
```
Comment (pwittrock) [inline kyaml/yaml/types.go:739]: nit: ConvertJSONToYAMLString

File: cmd/config/internal/commands/source_test.go (hunk 7 above)
Comment (pwittrock) [reply cmd/config/internal/commands/source_test.go:24]: Also update `sink_test.go` to make sure the results are written back as json

File: kyaml/kio/byteio_reader.go (hunk 4 above)
Comment (phanimarupaka) [reply kyaml/kio/byteio_reader.go]: Done.

Comment (phanimarupaka) [reply kyaml/kio/byteio_reader.go]: Done.

File: kyaml/kio/byteio_reader.go (hunk 6 above)
Comment (phanimarupaka) [reply kyaml/kio/byteio_reader.go:107]: I feel Path is more handy as it is the source of truth for the file ext and SerializationMode/DeserializationMode

File: cmd/config/internal/commands/source_test.go (hunk 7 above)
Comment (phanimarupaka) [reply cmd/config/internal/commands/source_test.go:24]: @pwittrock Updated sink_test.go, but it is adding single quotes to json string. Please verify if it's fine.

E2e test for source ? How would it be different from source_test.go

File: kyaml/yaml/types.go (hunk 8 above)
Comment (phanimarupaka) [reply kyaml/yaml/types.go:739]: Done

File: kyaml/kio/byteio_reader.go (hunk 6 above)
Comment (pwittrock) [reply kyaml/kio/byteio_reader.go:107]: What should path be if reading json from stdin?

File: cmd/config/internal/commands/source_test.go (hunk 7 above)
Comment (pwittrock) [reply cmd/config/internal/commands/source_test.go:24]: https://github.com/kubernetes-sigs/kustomize/pull/2552/commits/0e5e2648b345e847698791d3ae2fbf99679db73a#diff-8503fb052377a7f4a04632912645b0ef

Comment (pwittrock) [reply cmd/config/internal/commands/source_test.go:24]: It gives and extra level of broader coverage.  A single test should be fine, with sink_test.go covering most of the cases.

File: kyaml/kio/byteio_reader.go [hunk 9]
```diff
@@ -119,7 +123,20 @@ func (r *ByteReader) Read() ([]*yaml.RNode, error) {
 
 	index := 0
 	for i := range values {
-		decoder := yaml.NewDecoder(bytes.NewBufferString(values[i]))
+		value := values[i]
+
+		for _, g := range JSONMatch {
+			if match, err := filepath.Match(g, filepath.Ext(r.Path)); err != nil {
```
Comment (pwittrock) [inline kyaml/kio/byteio_reader.go:129]: why not just `g == filepath.Ext(r.Path)` ?

File: kyaml/yaml/types.go (hunk 8 above)
Comment (pwittrock) [reply kyaml/yaml/types.go:739]: The changes don't seem to be pushed.  Its still `ConvertJSONToYamlString`

File: kyaml/yaml/types.go [hunk 10]
```diff
@@ -724,20 +724,43 @@ func (rn *RNode) UnmarshalJSON(b []byte) error {
...
 	if err != nil {
-		return nil, err
+		return "", err
 	}
-	node, err := Parse(string(yml))
+	return string(yml), nil
+}
+
+// ConvertYamlNodeToJSONString returns json string from input yaml RNode
+func ConvertYamlNodeToJSONString(node *RNode) (string, error) {
```
Comment (pwittrock) [inline kyaml/yaml/types.go:753]: Same comment here

File: cmd/config/internal/commands/source_test.go (hunk 7 above)
Comment (phanimarupaka) [reply cmd/config/internal/commands/source_test.go:24]: Sure. I will add the e2e tests after your PR is merged so that I can reuse the methods. We can create a separate story and add them for all other commands as well.

File: cmd/config/internal/commands/source.go (hunk 5 above)
Comment (phanimarupaka) [reply cmd/config/internal/commands/source.go:74]: @pwittrock Does `fn run` depend on the file path annotation  ? As long as it doesn't depend , it should not have an effect IMO. The resource list is in yaml format only.

File: kyaml/kio/byteio_reader.go (hunk 6 above)
Comment (phanimarupaka) [reply kyaml/kio/byteio_reader.go:108]: @pwittrock It would be empty. Updated the comment.

Should we accept json input in stdin as well ? If so, we should parse the json prior to determine if it is json. The issue filed is for accepting json format file. Let me know your thoughts.

File: kyaml/yaml/types.go (hunk 8 above)
Comment (phanimarupaka) [reply kyaml/yaml/types.go:739]: It did. But it doesn't reflect here. Please visit Files Changed tab.

File: kyaml/kio/byteio_reader.go (hunk 9 above)
Comment (phanimarupaka) [reply kyaml/kio/byteio_reader.go]: As filepath.Ext(r.Path) is .json and g is *.json.

File: kyaml/kio/byteio_reader.go (hunk 6 above)
Comment (pwittrock) [reply kyaml/kio/byteio_reader.go:108]: The ByteReader shouldn't need to know or care about what the source of the bytes were -- whether it was a file, stdin, http request, etc.  Doing so breaks the abstraction -- it reads bytes.  If we would to make an input source that read from the apiserver, it would support reading json and use the ByteReader to parse the response bytes.

The format should also probably be an annotation that is set on the resource when it is read -- similar to how the index is an annotation on the resource.  This way the format would be preserved across processes.

File: kyaml/yaml/types.go (hunk 8 above)
Comment (pwittrock) [reply kyaml/yaml/types.go:739]: I am in the files changed tab and don't see it.

File: kyaml/kio/byteio_writer.go [hunk 11]
```diff
@@ -125,7 +128,29 @@ func (w ByteWriter) Write(nodes []*yaml.RNode) error {
...
+	rNode.SetYNode(doc)
+	err := w.encode(encoder, &rNode)
 	yaml.UndoSerializationHacksOnNodes(nodes)
 	return err
 }
+
+// encode encodes the input RNode to appropriate node format based on file path extension
+func (w ByteWriter) encode(encoder *yaml.Encoder, node *yaml.RNode) error {
+	for _, g := range JSONMatch {
+		if match, err := filepath.Match(g, filepath.Ext(w.Path)); err != nil {
```
Comment (pwittrock) [inline kyaml/kio/byteio_writer.go:141]: This seems redundant.  Either `filepath.Match(g, w.Path)` or `filepath.Ext(g) == filepath.Ext(w.Path)` seem simpler.

File: kyaml/kio/byteio_writer.go [hunk 12]
```diff
@@ -125,7 +128,29 @@ func (w ByteWriter) Write(nodes []*yaml.RNode) error {
...
+			return errors.Wrap(err)
+		} else if match {
```
Comment (pwittrock) [inline kyaml/kio/byteio_writer.go:143]: simply the code paths -- 

```go
if !match {
  continue
}

File: cmd/config/internal/commands/sink_test.go [hunk 13]
```diff
@@ -33,7 +33,7 @@ items:
     annotations:
       app: nginx2
       config.kubernetes.io/index: '0'
-      config.kubernetes.io/path: 'f1.yaml'
+      config.kubernetes.io/path: 'f1.json'
```
Comment (pwittrock) [inline cmd/config/internal/commands/sink_test.go:36]: rather than changing existing tests, would you create new ones specifically for json?

File: cmd/config/internal/commands/source.go (hunk 5 above)
Comment (pwittrock) [reply cmd/config/internal/commands/source.go:74]: Would you add a test to confirm that `run DIR/` and `source | run ` produce the same result when there are json inputs?

File: kyaml/kio/byteio_reader.go [hunk 14]
```diff
@@ -101,6 +102,10 @@ type ByteReader struct {
 	// WrappingKind is set by Read(), and is the kind of the object that
 	// the read objects were originally wrapped in.
 	WrappingKind string
+
+	// Path indicates file path which is being read
+	// empty if it is being read from stdin
+	Path string
```
Comment (pwittrock) [inline kyaml/kio/byteio_reader.go:108]: Can we detect the format from the input -- i.e. if the input is JSON parse the JSON, if the input is YAML parse the YAML -- in fact json is a subset of yaml, why do we need to know if this is json at all?

File: kyaml/kio/byteio_reader.go [hunk 15]
```diff
@@ -119,7 +124,22 @@ func (r *ByteReader) Read() ([]*yaml.RNode, error) {
...
 	for i := range values {
-		decoder := yaml.NewDecoder(bytes.NewBufferString(values[i]))
+		value := values[i]
+
+		if r.Path != "" {
+			for _, g := range JSONMatch {
+				if match, err := filepath.Match(g, filepath.Ext(r.Path)); err != nil {
+					return nil, errors.Wrap(err)
+				} else if match {
+					value, err = yaml.ConvertJSONToYamlString(value)
```
Comment (pwittrock) [inline kyaml/kio/byteio_reader.go:134]: Is this necessary -- won't the yaml parser parse json?

File: kyaml/kio/byteio_writer.go [hunk 16]
```diff
@@ -42,6 +43,9 @@ type ByteWriter struct {
 
 	// Sort if set, will cause ByteWriter to sort the the nodes before writing them.
 	Sort bool
+
+	// Path is the output file path to write to
+	Path string
```
Comment (pwittrock) [inline kyaml/kio/byteio_writer.go:48]: Same comment as above -- ByteReader shouldn't know about files.

File: kyaml/kio/byteio_writer.go (hunk 11 above)
Comment (pwittrock) [inline kyaml/kio/byteio_writer.go:141]: Rather than doing this -- lets just check for the annotation on the resource and if it says `config.k8s.io/format: json`, then write it as json.

File: kyaml/kio/byteio_writer.go [hunk 17]
```diff
@@ -90,9 +94,8 @@ func (w ByteWriter) Write(nodes []*yaml.RNode) error {
 	// don't wrap the elements
 	if w.WrappingKind == "" {
 		for i := range nodes {
-			err := encoder.Encode(nodes[i].Document())
-			if err != nil {
-				return errors.Wrap(err)
+			if err := w.encode(encoder, nodes[i]); err != nil {
```
Comment (pwittrock) [inline kyaml/kio/byteio_writer.go:97]: Would you write a unit test for this to make sure a collection of Nodes gets properly serialized as json.

e.g.

```json
{
  {...},
  {...}
}
```

File: kyaml/kio/byteio_writer.go [hunk 18]
```diff
@@ -125,7 +128,29 @@ func (w ByteWriter) Write(nodes []*yaml.RNode) error {
...
 	return err
 }
+
+// encode encodes the input RNode to appropriate node format based on file path extension
+func (w ByteWriter) encode(encoder *yaml.Encoder, node *yaml.RNode) error {
+	for _, g := range JSONMatch {
+		if match, err := filepath.Match(g, filepath.Ext(w.Path)); err != nil {
+			return errors.Wrap(err)
+		} else if match {
+			json, err := yaml.ConvertYamlNodeToJSONString(node)
```
Comment (pwittrock) [inline kyaml/kio/byteio_writer.go:144]: I'm not sure encoding each node as JSON individually works -- the whole collection needs to a be a json list right?  Or maybe the list needs to have exactly 1 item?  Yaml can contain multiple entries in a single byte stream without them being in a list, but JSON cannot.

This is going to be a bit more complicated to detect the right behavior.

**Comment by pwittrock** [comment]
If we want to make this simpler, it should be possible to just do the read bit and only change the glob for reading files (not the serialization).

Follow up with the write piece which is going to be more complicated.

File: kyaml/kio/byteio_writer.go (hunk 18 above)
Comment (pwittrock) [reply kyaml/kio/byteio_writer.go:144]: Probably error if there is more than 1 item and encoding is json

File: kyaml/kio/byteio_reader.go (hunk 6 above)
Comment (phanimarupaka) [reply kyaml/kio/byteio_reader.go]: Agree. Added AcceptJSON flag instead. PTAL

File: kyaml/yaml/types.go (hunk 8 above)
Comment (phanimarupaka) [reply kyaml/yaml/types.go]: Done.

File: kyaml/kio/byteio_reader.go (hunk 9 above)
Comment (phanimarupaka) [reply kyaml/kio/byteio_reader.go]: Removed the filematch logic.

File: kyaml/kio/byteio_writer.go (hunk 11 above)
Comment (phanimarupaka) [reply kyaml/kio/byteio_writer.go]: Done.

File: cmd/config/internal/commands/sink_test.go (hunk 13 above)
Comment (phanimarupaka) [reply cmd/config/internal/commands/sink_test.go]: @pwittrock Done.

File: kyaml/kio/byteio_reader.go (hunk 15 above)
Comment (phanimarupaka) [reply kyaml/kio/byteio_reader.go]: I tested it and it doesn't parse as expected. It adds escape characters etc. This way is more cleaner.

File: kyaml/kio/byteio_writer.go (hunk 16 above)
Comment (phanimarupaka) [reply kyaml/kio/byteio_writer.go]: Done.

File: kyaml/kio/byteio_writer.go (hunk 11 above)
Comment (phanimarupaka) [reply kyaml/kio/byteio_writer.go]: Done.

File: kyaml/kio/byteio_writer.go (hunk 18 above)
Comment (phanimarupaka) [reply kyaml/kio/byteio_writer.go]: Done. Added unit test to throw error if there are multiple json items.

File: cmd/config/internal/commands/source.go (hunk 5 above)
Comment (phanimarupaka) [reply cmd/config/internal/commands/source.go:74]: Planning to add an e2e test something like 

`kpt fn source myDir1/ | kpt fn run --image gcr.io/kpt-functions/label-namespace -- label_name=color label_value=orange | kpt fn sink myDir2/`

But it would need to pull an image. Is that fine and if so what is the correct place to do it. I tested manually and it works with json files.

Comment (phanimarupaka) [reply cmd/config/internal/commands/source.go:74]: @pwittrock We should have e2e tests something like (do we already have such explicit e2e test for functions involving source, run and sink)

`config fn source /myDir1 | config fn run --image gcr.io/kpt-functions/label-namespace -- label_name=color label_value=orange | config fn sink /myDir2`

which involves pulling down an image. Let me know if this is fine so I can add. I tested manually on local with json files and it works fine.

Comment (pwittrock) [reply cmd/config/internal/commands/source.go:74]: Yes -- but the test doesn't pull down an image -- it builds the image and uses the local copy

https://github.com/kubernetes-sigs/kustomize/blob/master/cmd/config/internal/commands/e2e/e2e_test.go#L747

Adding a test to that package SGTM

File: kyaml/kio/byteio_reader.go [hunk 19]
```diff
@@ -101,8 +101,16 @@ type ByteReader struct {
 	// WrappingKind is set by Read(), and is the kind of the object that
 	// the read objects were originally wrapped in.
 	WrappingKind string
+
+	// AcceptJSON indicates if the input json bytes should be processed
+	AcceptJSON bool
```
Comment (pwittrock) [inline kyaml/kio/byteio_reader.go:106]: Why is this necessary?

File: kyaml/kio/byteio_reader.go [hunk 20]
```diff
@@ -115,11 +123,32 @@ func (r *ByteReader) Read() ([]*yaml.RNode, error) {
...
 		return nil, errors.Wrap(err)
 	}
-	values := strings.Split(input.String(), "\n---\n")
+	inputStr := input.String()
+
+	// check if json is accepted and if input bytes are in json format
+	if r.AcceptJSON && yaml.ValidJSON(input.String()) {
+		// convert json to yaml string to generate resource list object
+		// with appropriate format annotation
+		yamlValue, err := yaml.ConvertJSONToYAMLString(input.String())
```
Comment (pwittrock) [inline kyaml/kio/byteio_reader.go:132]: Do we need this?  What happens if just let the YAML parser parse the JSON?

File: kyaml/kio/byteio_writer.go [hunk 21]
```diff
@@ -125,7 +125,29 @@ func (w ByteWriter) Write(nodes []*yaml.RNode) error {
...
 	yaml.UndoSerializationHacksOnNodes(nodes)
 	return err
 }
+
+// encode encodes the input RNode to appropriate node format based on file path extension
+func (w ByteWriter) encode(encoder *yaml.Encoder, node *yaml.RNode) error {
+	_, _, format, err := kioutil.GetFileAnnotations(node)
+	if err == nil && format == JSON {
+		je := json.NewEncoder(w.Writer)
+		if !w.KeepReaderAnnotations {
```
Comment (pwittrock) [inline kyaml/kio/byteio_writer.go:140]: why not do this before the if?

File: kyaml/kio/byteio_writer.go [hunk 22]
```diff
@@ -125,7 +125,29 @@ func (w ByteWriter) Write(nodes []*yaml.RNode) error {
...
+			_, err := node.Pipe(yaml.ClearAnnotation(kioutil.FormatAnnotation))
+			if err != nil {
+				return errors.Wrap(err)
+			}
+		}
+		err := je.Encode(node)
```
Comment (pwittrock) [inline kyaml/kio/byteio_writer.go:146]: What about?

```go
return errors.Wrap(je.Encode(node))
```

File: cmd/config/internal/commands/sink_test.go [hunk 23]
```diff
@@ -136,6 +136,46 @@ spec:
...
+	r.Command.SetArgs([]string{d})
+	if !assert.NoError(t, r.Command.Execute()) {
+		t.FailNow()
+	}
+
+	actual, err := ioutil.ReadFile(filepath.Join(d, "f1.json"))
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+	expected := `{"kind":"Deployment","metadata":{"annotations":{"app":"nginx2"},"labels":{"app":"nginx2"},"name":"foo"},"spec":{"replicas":1}}
```
Comment (pwittrock) [inline cmd/config/internal/commands/sink_test.go:172]: We probably want to output to be multiline.

File: kyaml/yaml/types.go [hunk 24]
```diff
@@ -722,22 +722,36 @@ func (rn *RNode) UnmarshalJSON(b []byte) error {
...
-	return node, nil
+	yml, err := yaml.Marshal(body)
+	if err != nil {
+		return "", err
+	}
+	return string(yml), nil
+}
+
+func ValidJSON(s string) bool {
+	var js map[string]interface{}
```
Comment (pwittrock) [inline kyaml/yaml/types.go:753]: Why not?

```go
func IsJSON(str string) bool {
    var js json.RawMessage
    return json.Unmarshal([]byte(str), &js) == nil
}
```

https://stackoverflow.com/questions/22128282/how-to-check-string-is-in-json-format

Comment (phanimarupaka) [reply kyaml/yaml/types.go]: Done.

File: cmd/config/internal/commands/sink_test.go (hunk 23 above)
Comment (phanimarupaka) [reply cmd/config/internal/commands/sink_test.go]: Done.

File: kyaml/kio/byteio_writer.go (hunk 22 above)
Comment (phanimarupaka) [reply kyaml/kio/byteio_writer.go]: Done.

File: kyaml/kio/byteio_writer.go (hunk 21 above)
Comment (phanimarupaka) [reply kyaml/kio/byteio_writer.go]: Done.

File: kyaml/kio/byteio_reader.go (hunk 19 above)
Comment (phanimarupaka) [reply kyaml/kio/byteio_reader.go:107]: To reduce the blast radius, wanted to introduce it only to source command. We can get rid of it eventually when we want to roll out to all commands.

File: kyaml/kio/byteio_reader.go (hunk 20 above)
Comment (phanimarupaka) [reply kyaml/kio/byteio_reader.go:133]: I tested it and it is not parsing json as expected.

File: kyaml/kio/byteio_reader.go [hunk 25]
```diff
@@ -115,11 +123,32 @@ func (r *ByteReader) Read() ([]*yaml.RNode, error) {
...
+		if err != nil {
+			return nil, err
+		}
+		if r.SetAnnotations == nil {
+			r.SetAnnotations = map[string]string{}
+		}
+		if !r.OmitReaderAnnotations {
+			r.SetAnnotations[kioutil.FormatAnnotation] = JSON
```
[the remark this answers is not in the document]
Comment (phanimarupaka) [reply kyaml/kio/byteio_reader.go:141]: @pwittrock How about the stdin ? This place is common for both so that format can be set here.

File: kyaml/kio/byteio_reader.go (hunk 20 above)
Comment (pwittrock) [reply kyaml/kio/byteio_reader.go:133]: Weird.  What is happening?

File: kyaml/kio/byteio_reader.go (hunk 15 above)
Comment (pwittrock) [reply kyaml/kio/byteio_reader.go]: What do you mean it adds escape characters?

https://github.com/kubernetes-sigs/kustomize/pull/2575

Comment (phanimarupaka) [reply kyaml/kio/byteio_reader.go]: @pwittrock It parses standalone json node but it has problem while wrapping it in a ResourceList. The output looks like this.

```
apiVersion: config.kubernetes.io/v1alpha1
kind: ResourceList
items:
- {"kind": "Deployment", "metadata": {"labels": {"app": "nginx2"}, "name": "foo",
    "annotations": {"app": "nginx2", config.kubernetes.io/index: '0', config.kubernetes.io/path: 'f1.json'}},
  "spec": {"replicas": 1}}
```

Instead, if we convert it to yaml string, rest of it will be taken care by existing logic. WDYT?

File: cmd/config/internal/commands/source.go (hunk 5 above)
Comment (phanimarupaka) [reply cmd/config/internal/commands/source.go:74]: @pwittrock added a test.

**Comment by pwittrock** [comment]
@phanimarupaka if you don't mind, I'd like to think more about the complexity this is introducing and if there is a simpler way to meet the immediate goals.

What json wasn't being parsed by the yaml parser?

**Comment by phanimarupaka** [comment]
@pwittrock Here is the comment. TL;DR json is being parsed fine by yaml parser. So the Convert method is removed now. https://github.com/kubernetes-sigs/kustomize/pull/2546#discussion_r437149753

Let me know what other parts are complex so that we can simplify. I agree that this is core piece of kyaml and should be accurate.

**Comment by phanimarupaka** [comment]
@pwittrock I have simplified the logic. The only change new method added is to prettify json.

**Comment by pwittrock** [comment]
@phanimarupaka Thanks, looks great.  I really like how simple this is.
